### PR TITLE
[SPARK-29369][SQL] Support string intervals without the `interval` prefix

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to move interval parsing to `CalendarInterval.fromCaseInsensitiveString()` which throws an `IllegalArgumentException` for invalid strings, and reuse it from `CalendarInterval.fromString()`. The former one handles `IllegalArgumentException` only and returns `NULL` for invalid interval strings. This will allow to support interval strings without the `interval` prefix in casting strings to intervals and in interval type constructor because they use `fromString()` for parsing string intervals.

For example:
```sql
spark-sql> select cast('1 year 10 days' as interval);
interval 1 years 1 weeks 3 days
spark-sql> SELECT INTERVAL '1 YEAR 10 DAYS';
interval 1 years 1 weeks 3 days
```

### Why are the changes needed?
To maintain feature parity with PostgreSQL which supports interval strings without prefix:
```sql
# select interval '2 months 1 microsecond';
        interval        
------------------------
 2 mons 00:00:00.000001
```
and to improve Spark SQL UX.

### Does this PR introduce any user-facing change?
Yes, previously parsing of interval strings without `interval` gives `NULL`:
```sql
spark-sql> select interval '2 months 1 microsecond';
NULL
```
After:
```sql
spark-sql> select interval '2 months 1 microsecond';
interval 2 months 1 microseconds
```

### How was this patch tested?
- Added new tests to `CalendarIntervalSuite.java`
- A test for casting strings to intervals in `CastSuite`
- Test for interval type constructor from strings in `ExpressionParserSuite`

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @dongjoon-hyun @zsxwing May I ask you to take a look at this PR.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java [hunk 1]
```diff
@@ -73,45 +72,50 @@ private static long toLong(String s) {
...
 
   /**
-   * Convert a string to CalendarInterval. Unlike fromString, this method can handle
+   * Convert a string to CalendarInterval. This method can handle
    * strings without the `interval` prefix and throws IllegalArgumentException
    * when the input string is not a valid interval.
    *
    * @throws IllegalArgumentException if the string is not a valid internal.
    */
   public static CalendarInterval fromCaseInsensitiveString(String s) {
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: Is it the only place we parse interval string? I thought we parse it with antlr parser.

Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: antlr parser does this as well but it parses sql elements like
```sql
spark-sql> select interval 10 days 1 second;
interval 1 weeks 3 days 1 seconds
```
here is only the place where we parse string values:
```sql
spark-sql> select interval 'interval 10 days 1 second';
interval 1 weeks 3 days 1 seconds
```

Comment (cloud-fan) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: This looks duplicated. Shall we add a `parseInterval` method to the `ParserInterface` interface and call the parser here?

Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: Maybe something has been duplicated, and can be reused but this is heavy refactoring for this PR. 

For instance, `AstBuilder.visitInterval` gets already split interval units but `CalendarInterval.fromString()` uses regular expression to parse & split: https://github.com/apache/spark/blob/b10344956d672fca495d0684b351ab5cf9f210d8/common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java#L50-L53

If you don't mind, I would try to do that in a separate PR.

Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: And your regexp is not tolerant to the order of interval units, see:
```sql
spark-sql> select interval 'interval 1 microsecond 2 months';
NULL
spark-sql> select interval 1 microsecond 2 months;
interval 2 months 1 microseconds
```

Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: Let's keep them separate so far. And I will try to write flexible and common code for parsing string interval that could handle other features found in https://github.com/apache/spark/pull/26055

Comment (cloud-fan) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: SGTM

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java [hunk 2]
```diff
@@ -73,45 +72,50 @@ private static long toLong(String s) {
...
-      sInLowerCase.startsWith("interval ") ? sInLowerCase : "interval " + sInLowerCase;
-    CalendarInterval cal = fromString(interval);
-    if (cal == null) {
+    String trimmed = s.trim();
+    if (trimmed.isEmpty()) {
+      throw new IllegalArgumentException("Interval cannot be blank");
+    }
+    String prefix = "interval";
+    String intervalStr = trimmed;
+    if (!intervalStr.regionMatches(true, 0, prefix, 0, prefix.length())) {
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: I added comments about this.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java [hunk 3]
```diff
@@ -73,45 +72,53 @@ private static long toLong(String s) {
...
+    String trimmed = s.trim();
+    if (trimmed.isEmpty()) {
+      throw new IllegalArgumentException("Interval cannot be blank");
+    }
+    String prefix = "interval";
+    String intervalStr = trimmed;
+    // Checks the given interval string does not start with the `interval` prefix
+    if (!intervalStr.regionMatches(true, 0, prefix, 0, prefix.length())) {
+      // Prepend `interval` if it does not present because
+      // the regular expression strictly require it.
```
Comment (MaxGekk) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: I have not figure out how to modify the regular expression to make the `interval` prefix optional.

Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: Probably, this needs this feature https://www.regular-expressions.info/branchreset.html which Java's regexps doesn't have.

Comment (cloud-fan) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: How about something like
```
String intervalStr = trimmed.toLowerCase();
if (intervalStr.startsWith("interval")) {
  intervalStr = intervalStr.drop(8)
}
```

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@cloud-fan @dongjoon-hyun @srowen @HyukjinKwon May I ask you to have a look at the PR.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java [hunk 4]
```diff
@@ -73,45 +72,53 @@ private static long toLong(String s) {
...
-      sInLowerCase.startsWith("interval ") ? sInLowerCase : "interval " + sInLowerCase;
-    CalendarInterval cal = fromString(interval);
-    if (cal == null) {
+    String trimmed = s.trim();
+    if (trimmed.isEmpty()) {
+      throw new IllegalArgumentException("Interval cannot be blank");
+    }
+    String prefix = "interval";
+    String intervalStr = trimmed;
+    // Checks the given interval string does not start with the `interval` prefix
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: Why not just call `trimmed.toLowerCase.startsWith("interval")`? For perf reasons?

Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: Yes, don't want to lower case the **entire string** and allocate memory for new one to only compare small prefix.

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java (hunk 3 above)
Comment (srowen) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: Can you just start the regex with `(interval)?`? then the first matching group is either null or "interval", and the rest should match the same way?

Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: As far as I remember I tried this regex, and it didn't work. Have you tried it?

Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: I have checked, it doesn't work:
```scala
scala> def unitRegex(unit: String) = "(?:\\s+(-?\\d+)\\s+" + unit + "s?)?"
unitRegex: (unit: String)String

scala> val p = Pattern.compile("(interval)?" + unitRegex("year") + unitRegex("month") +
     |     unitRegex("week") + unitRegex("day") + unitRegex("hour") + unitRegex("minute") +
     |     unitRegex("second") + unitRegex("millisecond") + unitRegex("microsecond"),
     |     Pattern.CASE_INSENSITIVE)
p: java.util.regex.Pattern = (interval)?(?:\s+(-?\d+)\s+years?)?(?:\s+(-?\d+)\s+months?)?(?:\s+(-?\d+)\s+weeks?)?(?:\s+(-?\d+)\s+days?)?(?:\s+(-?\d+)\s+hours?)?(?:\s+(-?\d+)\s+minutes?)?(?:\s+(-?\d+)\s+seconds?)?(?:\s+(-?\d+)\s+milliseconds?)?(?:\s+(-?\d+)\s+microseconds?)?

scala> val m = p.matcher("1 month 1 second")
m: java.util.regex.Matcher = java.util.regex.Matcher[pattern=(interval)?(?:\s+(-?\d+)\s+years?)?(?:\s+(-?\d+)\s+months?)?(?:\s+(-?\d+)\s+weeks?)?(?:\s+(-?\d+)\s+days?)?(?:\s+(-?\d+)\s+hours?)?(?:\s+(-?\d+)\s+minutes?)?(?:\s+(-?\d+)\s+seconds?)?(?:\s+(-?\d+)\s+milliseconds?)?(?:\s+(-?\d+)\s+microseconds?)? region=0,16 lastmatch=]

scala> m.matches()
res0: Boolean = false
```

Comment (srowen) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: I tried simply `"(interval)?(.+)".r` and it worked as expected on inputs like `"abc"` and `"interval abc"`. It's a toy example and not sure if it interacts unexpectedly with the rest of the matching. no big deal, just leave it.

**Pull request closed without merging**